README.md

```diff
@@ -285,6 +285,61 @@ There are many places in the format for compatible extensions:
 - Encodings: Encodings are specified by enum and more can be added in the future.
 - Page types: Additional page types can be added and safely skipped.
 
+### Thrift extensions
+Thrift is used for metadata. The Thrift spec mandates that unknown fields are
+skipped. To facilitate extensions Parquet reserves field-id `32767` of *every*
+struct as an ignorable extension point. More specifically Parquet guarantees
+that field-id `32767` will *never* be seen in the official Thrift IDL. The type
+of this field is always `binary`. 
+
+Such extensions can easily be appended to an existing Thrift serialized message
+without any special APIs. Sample `C++` implementation is provided:
+
+```c++
+std::string AppendExtension(std::string thrift, std::string ext) {
+  auto append_uleb = [](uint32_t x, std::string* out) {
+    while (true) {
+      int c = x & 0x7F;
+      if ((x >>= 7) == 0) {
+        out->push_back(c);
+        return;
+      } else {
+        out->push_back(c | 0x80);
+      }
+    }
+  };
+  thrift.pop_back();  // remove the trailing 0
+  thrift += "\x08\xFF\xFF\x01";
+  append_uleb(ext.size(), &thrift);
+  thrift += ext;
+  thrift += "\x00";  // add the trailing 0 back
+  return thrift;
+}
+```
+
+Additionally the binary extension MUST have a specific form in order to be
+unambiguously identifiable by parsers that know of it and, as a corollary,
+impossible to be accidentally generated by user data.
+
+```
+N bytes: the extension data
```

Comment:
I'd be tempted to declare that the first 2 bytes map to a format ID. This does stop extension formats being completely closed, provided the authors don't change it and document the structure

Reply:
@steveloughran could you elaborate I'm don't think I understand what you  are suggesting?

Reply:
In practice they will start with some ID because vendors will experiment and need to disambiguate. But I do not want to specify this since it won't make it in the official spec.

For example, Databricks experiments with footers. It uses the above mechanism and rolls this out to several customers. The first byte contains some id for the blob:
1. some custom encoding to speed up colchunk offset parsing only, with the idea that this is enough to pipeline data I/O with metadata parsing
2. some more refinement to the encoding
3. another experiment, now with more structured data, say flatbuffers
4. refinement of (3)
5. etc
6. Final version proposed for inclusion to the spec
At this point (6) does not have to end up in the official spec. The way it will be added is either through changing the magic of parquet, or through adding a new binary field in `FileMetaData` that contains the final version of the encoding.

Reply:
i'm just thinking "if ever you want to do any kind of reading across other apps, having a hint of structure would be good". Even if it is just to say "this file has databricks "type 2" footers rather than "something by databricks"

Reply:
I am still tempted to leave this out as I find it of little utility and unnecessary constraint. What if the designer actually uses some id or version but that's better put at the end of the binary blob before the crc of the len? Why do we want to force it to the start?

I am adding a paragraph to strongly suggest that implementors add an `id` to their extension to make interoperability easier in case it is needed but I leave it up to them to incorporate where they deem most appropriate.

Reply:
After thinking some more, I simplified the specification further.

I have removed all the fields from the encoding itself except the magic 3 byte disambiguator.

The rest are moved into an example section that describes how one might design a `FileMetaData` replacement.

The extension mechanism is very flexible, one might extend `ColumnMetaData` with some extra information and in that design the extension may be very different than the `FileMetaData` replacement.